packages/clerk-js/src/ui/hooks/usePasswordComplexity.ts

```diff
@@ -0,0 +1,126 @@
+import type { PasswordSettingsData } from '@clerk/types';
+import { useCallback, useMemo, useState } from 'react';
+
+import { localizationKeys, useLocalizations } from '../localization';
+
+const testComplexityCases = (
+  password: string,
+  {
+    minLength,
+    maxLength,
+  }: {
+    minLength: number;
+    maxLength: number;
+  },
+) => {
+  return {
+    max_length: password.length < maxLength,
+    min_length: password.length >= minLength,
+    require_numbers: /\d/.test(password),
+    require_lowercase: /[a-z]/.test(password),
+    require_uppercase: /[A-Z]/.test(password),
+    require_special_char: /[!"#$%&'()*+,-./:;<=>?@[\]^_`{|}~]/.test(password),
+  };
+};
+
+type ComplexityErrorMessages = {
+  [key in keyof Partial<Omit<PasswordSettingsData, 'disable_hibp' | 'min_zxcvbn_strength' | 'show_zxcvbn'>>]: string;
+};
+
+export const usePasswordComplexity = (
+  config: Omit<PasswordSettingsData, 'disable_hibp' | 'min_zxcvbn_strength' | 'show_zxcvbn'>,
```

Comment:
Nit: In general I feel that keeping the types and the function inputs separated is more readable.
Eg:
```ts
type UsePasswordComplexityConfig = ...
type UsePasswordComplexityCbs = ...
type UsePasswordComplexity = (config: UsePasswordComplexityConfig, cbs: UsePasswordComplexityCbs) => void;

export const usePasswordComplexity: UsePasswordComplexity = (config, callbacks) => {
  const { onValidationFailed = noop, onValidationSuccess = noop } = callbacks || {};
  ...
}
```